### DISCUSSION TITLE
0.23.0-rc2: Routes conflict

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#5615] [Q&A] [answered]
How to Golang std mux routing with static directory server.  In Echo this used to be something like a fallback, but it conflicts with the new Golang 1.22 mux features.  I can't get a dynamic root page without needing to specific many static paths.

```js
routerAdd("get", "/", (e) => {
  return e.string(200, "YES!")
})

routerAdd("get", "/{path...}", $apis.static(`${__hooks}/../public`, false))
```

**Comment by ganigeorgiev** [comment]
What was the old version looks like (_ideally with example matching paths_)?

Could you elaborate what actually is the intended behavior of the fallback route?

**Reply by ganigeorgiev** [reply]
If you want a custom 404/error route, you can try to catch the `e.next()` error from a middleware, something like:

```js
routerUse((e) => {
    try {
        e.next()
    } catch(err) {
        // edit: e.status() will not work in this case because the response is not written yet so we use "err.value" to access the wrapped goja ApiError
        // edit2: this middleware affects all routes so you may want to do some further checks based on the e.request fields
        if (err.value.status == 404) {
            return e.string(200, "YES!") 
        } else {
            // edit3: rethrow in case the error doesn't match your criteria
            throw err
        }
    }
})
```

**Reply by dschissler** [reply]
This used to allow the root path (not path in browser URL bar) to be this programmatic route, while files where then served as a fallback if they existed.

```js
routerAdd("get", "/", (c) => {
  const {camCfgs} = require(`${__hooks}/cams.js`)

  const html = require(`${__hooks}/template.js`)(
    `layouts/layout`,
    `home`, 
    `partials/webcam`,
  ).render({
    'webcams': camCfgs.map(({src, color}) => {
      return {
        src,
        color,
      }
    })
  })
  return c.html(200, html)
})

routerAdd("get", "/*", $apis.staticDirectoryHandler(`${__hooks}/../public`, false))

```

**Reply by ganigeorgiev** [reply, marked as answer]
@dschissler So the fallback is intended to serve a custom root (index) page, right?

In that case appending `{$}` to the first route to indicate matching the exact ending should suffice, aka. the below routes:

```js
// route1
routerAdd("get", "/{$}", (e) => {
  return e.string(200, "YES!")
})

// route2
routerAdd("get", "/{path...}", $apis.static(`${__hooks}/../public`, false))
```
should match the following paths:

- `/` - route1
- `/anything` - route2

**Reply by ganigeorgiev** [reply]
I think maybe the confusion comes from the fact that with the std Go mux, patterns ending with slash will match any routes beginning with that path.

I'll make sure to add it in the new documentation, but you can find more information about the Go std mux pattern matching in https://pkg.go.dev/net/http#ServeMux and https://go.dev/blog/routing-enhancements.

**Reply by dschissler** [reply]
You're solution works.  I was trying `/$` and that sort of thing.  I'm perfectly fine learning this new builtin Golang magic though.
